elastic/kibana
# fix(slo): history details data

Resolves https://github.com/elastic/kibana/issues/182251

## 🌮  Summary

This PR fixes the calculation of the historical summary data when an ad-hoc date range is provided. Indeed, when querying the SLI data for a rolling SLO with a provided date range, we need to query since the start of the range minus the slo rolling window duration. Otherwise the cumulative sum is not correct.

One way to see it is that when we zoom into the chart, the values should not change (when keeping the same bucket sizes, i.e when the range duration is in the same order of magnitude than the time window duration)

I'm also cleaning up the range type, by keeping only one common type, e.g. using Date type. I think that string or number for representing dates are dangerous, and difficult to follow in a codebase. 

I also need to go back to the historical summary client and clean up the DateRange i've introduced to fix the issue aforementioned issue. it's not very clean, and pretty confusing at the moment. So far, I only came up with a comment. Any suggestion for improving this part of the code, `getDateRange()`? 

I've used this to validate the tests: https://docs.google.com/spreadsheets/d/1uDoqq4sJPt2Dnksc1U2x6zAfQB73SK2udvweLl6ODuo/edit#gid=908013864

## How to test

- Run data forge for 2 weeks: `node x-pack/scripts/data_forge.js --events-per-cycle 50 --lookback now-14d --dataset fake_stack --install-kibana-assets --kibana-url http://localhost:5601/kibana`
- Go get a coffee, it will take time
- Create an SLO over 7days rolling window (either Timeslices or Occurrences)
- When going to the history page without changing the time picker value, the historical charts should be the same than the overview page (look at the beginning of the chart)
- When zooming in with a date range > 3days (to keep the same bucket size interval for the calculations), any value should be equal to the previously shown value for the same date.


https://github.com/elastic/kibana/assets/1376800/5eed2560-f706-4910-8b92-97091c0c83a3
